File: java/flight/src/main/java/org/apache/arrow/flight/grpc/MetadataAdapter.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.arrow.flight.grpc;

import java.util.Set;

import org.apache.arrow.flight.CallHeaders;

import io.grpc.Metadata;
import io.grpc.Metadata.Key;

/**
 * A mutable adapter between the gRPC Metadata object and the Flight headers interface.
 *
 * <p>This allows us to present the headers (metadata) from gRPC without copying to/from our own object.
 */
class MetadataAdapter implements CallHeaders {

  private final Metadata metadata;

  MetadataAdapter(Metadata metadata) {
    this.metadata = metadata;
  }

  @Override
  public byte[] getBinary(String key) {
    return this.metadata.get(Key.of(key, Metadata.BINARY_BYTE_MARSHALLER));
  }

  @Override
  public String getText(String key) {
    return this.metadata.get(Key.of(key, Metadata.ASCII_STRING_MARSHALLER));
  }

  @Override
  public void putBinary(String key, byte[] value) {
    this.metadata.put(Key.of(key, Metadata.BINARY_BYTE_MARSHALLER), value);
  }

  @Override
  public void putText(String key, String value) {
    this.metadata.put(Key.of(key, Metadata.ASCII_STRING_MARSHALLER), value);
  }

  @Override
  public Set<String> keys() {
    return this.metadata.keys();
  }

  @Override
  public boolean containsKey(String key) {
    final Key<?> grpcKey = key.endsWith("-bin") ? Key.of(key, Metadata.BINARY_BYTE_MARSHALLER)
        : Key.of(key, Metadata.ASCII_STRING_MARSHALLER);
    return this.metadata.containsKey(grpcKey);
  }

  public String toString() {
    return this.metadata.toString();
  }
}

File: csharp/src/Apache.Arrow.Flight/Middleware/MetadataAdapter.cs
// Licensed to the Apache Software Foundation (ASF) under one or more
// contributor license agreements. See the NOTICE file distributed with
// this work for additional information regarding copyright ownership.
// The ASF licenses this file to You under the Apache License, Version 2.0
// (the "License"); you may not use this file except in compliance with
// the License.  You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System;
using System.Collections.Generic;
using System.Linq;
using Apache.Arrow.Flight.Middleware.Interfaces;
using Grpc.Core;

namespace Apache.Arrow.Flight.Middleware;

public class MetadataAdapter : ICallHeaders
{
    private readonly Metadata _metadata;

    public MetadataAdapter(Metadata metadata)
    {
        _metadata = metadata ?? throw new ArgumentNullException(nameof(metadata));
    }

    public string this[string key] => Get(key);

    public string Get(string key)
    {
        return _metadata.FirstOrDefault(e =>
            !e.IsBinary && e.Key.Equals(key, StringComparison.OrdinalIgnoreCase))?.Value;
    }

    public byte[] GetBytes(string key)
    {
        return _metadata.FirstOrDefault(e =>
            e.IsBinary && e.Key.Equals(NormalizeBinaryKey(key), StringComparison.OrdinalIgnoreCase))?.ValueBytes;
    }

    public IEnumerable<string> GetAll(string key)
    {
        return _metadata
            .Where(e => !e.IsBinary && e.Key.Equals(key, StringComparison.OrdinalIgnoreCase))
            .Select(e => e.Value);
    }

    public IEnumerable<byte[]> GetAllBytes(string key)
    {
        var binaryKey = NormalizeBinaryKey(key);
        return _metadata
            .Where(e => e.IsBinary && e.Key.Equals(binaryKey, StringComparison.OrdinalIgnoreCase))
            .Select(e => e.ValueBytes);
    }

    public void Insert(string key, string value)
    {
        _metadata.Add(key, value);
    }

    public void Insert(string key, byte[] value)
    {
        _metadata.Add(NormalizeBinaryKey(key), value);
    }

    public ISet<string> Keys =>
        new HashSet<string>(_metadata.Select(e =>
                e.IsBinary ? DenormalizeBinaryKey(e.Key) : e.Key),
            StringComparer.OrdinalIgnoreCase);

    public bool ContainsKey(string key)
    {
        return _metadata.Any(e =>
            e.Key.Equals(key, StringComparison.OrdinalIgnoreCase) ||
            e.Key.Equals(NormalizeBinaryKey(key), StringComparison.OrdinalIgnoreCase));
    }

    private static string NormalizeBinaryKey(string key)
        => key.EndsWith(Metadata.BinaryHeaderSuffix, StringComparison.OrdinalIgnoreCase)
            ? key
            : key + Metadata.BinaryHeaderSuffix;

    private static string DenormalizeBinaryKey(string key)
        => key.EndsWith(Metadata.BinaryHeaderSuffix, StringComparison.OrdinalIgnoreCase)
            ? key.Substring(0, key.Length - Metadata.BinaryHeaderSuffix.Length)
            : key;
}

public static class MetadataAdapterExtensions
{
    public static bool TryGet(this ICallHeaders headers, string key, out string value)
    {
        value = headers.Get(key);
        return value is not null;
    }
}